src/App.tsx

```diff
@@ -15,48 +14,125 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+interface State {
+  sortedGoods: string[];
+  showResetButton: boolean;
+  sortByAlphabeticallyActive?: boolean;
+  sortByLengthActive?: boolean;
+  reverseActive?: boolean;
+}
+
 export const App: React.FC = () => {
+  const initialState: State = {
+    sortedGoods: [...goodsFromServer],
+    showResetButton: false,
+  };
+
+  const [state, setState] = useState<State>(initialState);
+
+  const {
+    sortedGoods,
+    showResetButton,
+    sortByAlphabeticallyActive,
+    sortByLengthActive,
+    reverseActive,
+  } = state;
+
+  const setActiveState = (activeSort: string) => ({
+    sortByAlphabeticallyActive: activeSort === 'alphabetical',
+    sortByLengthActive: activeSort === 'length',
+    reverseActive: activeSort === 'reverse',
+  });
+
+  const updateState = (updatedState: Partial<State>) => {
+    setState(prevState => ({
+      ...prevState,
+      ...updatedState,
+    }));
+  };
+
+  const sortArray = (sortFunction: (a: string, b: string)
+  => number, activeSort: string) => {
+    const sortedArray = [...sortedGoods].sort(sortFunction);
+
+    updateState({
+      sortedGoods: sortedArray,
+      showResetButton: true,
+      ...setActiveState(activeSort),
+    });
+  };
+
+  const reverseArray = () => {
+    const reversedArray = [...sortedGoods].reverse();
+
+    updateState({
+      sortedGoods: reversedArray,
+      reverseActive: !reverseActive,
+    });
+  };
+
+  const resetArray = () => {
+    setState(initialState);
+  };
+
+  const sortAlphabetically = () => {
+    sortArray((a, b) => a.localeCompare(b), 'alphabetical');
+  };
+
+  const sortByLength = () => {
+    sortArray((a, b) => a.length - b.length, 'length');
+  };
+
+  const handleReverse = () => {
+    if (reverseActive) {
```

Comment:
The handleReverse function resets the array when 'reverseActive' is true. This could be misleading as it's name suggests it does reversal operation, not a reset. Consider renaming the function to reflect its actual functionality.